gwpopulation/hyperpe.py

```diff
@@ -115,13 +144,40 @@ def noise_log_likelihood(self):
     def log_likelihood(self):
         return self.noise_log_likelihood() + self.log_likelihood_ratio()
 
-    def _compute_per_event_ln_bayes_factors(self):
-        return -np.log(self.samples_per_posterior) + xp.log(
-            xp.sum(self.hyper_prior.prob(self.data) / self.sampling_prior, axis=-1)
-        )
+    def _pop_added(self, added_keys):
+        if added_keys is not None:
+            for key in added_keys:
+                self.parameters.pop(key)
 
-    def _get_selection_factor(self):
-        return -self.n_posteriors * xp.log(self.selection_function(self.parameters))
+    def _compute_per_event_ln_bayes_factors(self, return_uncertainty=True):
+        weights = self.hyper_prior.prob(self.data) / self.sampling_prior
+        expectation = xp.mean(weights, axis=-1)
+        if return_uncertainty:
+            square_expectation = xp.mean(weights**2, axis=-1)
+            variance = (square_expectation - expectation**2) / (
+                self.samples_per_posterior * square_expectation
+            )
+            return xp.log(expectation), variance
+        else:
+            return xp.log(expectation)
+
+    def _get_selection_factor(self, return_uncertainty=True):
+        selection, variance = self._selection_function_with_uncertainty()
+        total_selection = -self.n_posteriors * xp.log(selection)
+        if return_uncertainty:
+            total_variance = self.n_posteriors**2 * variance
```

Comment:
I think if you're doing it this way and want this function to return a value that you can add in quadrature to the other uncertainties, you'll want this line to look like
```suggestion
            total_variance = self.n_posteriors**2 * variance / selection**2
```

This is because we want the uncertainty in the total log likelihood, and dlnL/dlnPdet = 1/Pdet. Since the term `variance` here is the absolute variance in Pdet from the Monte Carlo sum, and we want to propagate this into the variance in the total log likelihood, the variance of the total log likelihood will go like the variance computed in 125 + var(Pdet)/Pdet**2

Reply:
[MC.pdf](https://github.com/ColmTalbot/gwpopulation/files/8006120/MC.pdf)
I couldn't figure out how to type good equations in the comments, so at the risk of overkill I attached this in LaTeX. Basically, I think the way you are calculating the variance of the total likelihood, by summing the contribution from Pdet in quadrature assumes that the variance from the Pdet term is a relative variance, whereas I believe it is in fact an absolute variance. i.e.  the variance in the Monte Carlo that you call variance in this line is equation 6 in the attached file. But when we propagate to statistical uncertainties in the total log likelihood, the contribution should be in the form of a relative variance as in equation 9. Basically I think what is being done now does not include the \mu^2 term in the denominator which I think it should